tfhe/src/core_crypto/keycache.rs

```diff
@@ -0,0 +1,279 @@
+        P::access(self).clear_in_memory_cache();
+    }
+}
+
+pub trait KeyCacheAccess: Serialize + DeserializeOwned + Copy + PartialEq + NamedParam {
+    type Keys: DeserializeOwned + Serialize;
+
+    fn access(keycache: &KeyCache) -> &KeyCacheCoreImpl<Self, Self::Keys>;
+}
+
+impl KeyCacheAccess for MultiBitTestParams<u32> {
+    type Keys = MultiBitBootstrapKeys<u32>;
+
+    fn access(keycache: &KeyCache) -> &KeyCacheCoreImpl<Self, Self::Keys> {
+        &keycache.u32_multi_bit_cache
+    }
+}
+
+impl KeyCacheAccess for MultiBitTestParams<u64> {
+    type Keys = MultiBitBootstrapKeys<u64>;
+
+    fn access(keycache: &KeyCache) -> &KeyCacheCoreImpl<Self, Self::Keys> {
+        &keycache.u64_multi_bit_cache
+    }
+}
+
+impl KeyCacheAccess for ClassicTestParams<u32> {
+    type Keys = ClassicBootstrapKeys<u32>;
+
+    fn access(keycache: &KeyCache) -> &KeyCacheCoreImpl<Self, Self::Keys> {
+        &keycache.u32_classic_cache
+    }
+}
+
+impl KeyCacheAccess for ClassicTestParams<u64> {
+    type Keys = ClassicBootstrapKeys<u64>;
+
+    fn access(keycache: &KeyCache) -> &KeyCacheCoreImpl<Self, Self::Keys> {
+        &keycache.u64_classic_cache
+    }
+}
+
+impl KeyCacheAccess for ClassicTestParams<u128> {
+    type Keys = ClassicBootstrapKeys<u128>;
+
+    fn access(keycache: &KeyCache) -> &KeyCacheCoreImpl<Self, Self::Keys> {
+        &keycache.u128_classic_cache
+    }
+}
+
+impl KeyCacheAccess for FftTestParams<u32> {
+    type Keys = FftBootstrapKeys<u32>;
+
+    fn access(keycache: &KeyCache) -> &KeyCacheCoreImpl<Self, Self::Keys> {
+        &keycache.u32_fft_cache
+    }
+}
+
+impl KeyCacheAccess for FftTestParams<u64> {
+    type Keys = FftBootstrapKeys<u64>;
+
+    fn access(keycache: &KeyCache) -> &KeyCacheCoreImpl<Self, Self::Keys> {
+        &keycache.u64_fft_cache
+    }
+}
+
+impl KeyCacheAccess for FftTestParams<u128> {
+    type Keys = FftBootstrapKeys<u128>;
+
+    fn access(keycache: &KeyCache) -> &KeyCacheCoreImpl<Self, Self::Keys> {
+        &keycache.u128_fft_cache
+    }
+}
+
+impl KeyCacheAccess for FftWopPbsTestParams<u64> {
+    type Keys = FftWopPbsKeys<u64>;
+
+    fn access(keycache: &KeyCache) -> &KeyCacheCoreImpl<Self, Self::Keys> {
+        &keycache.u64_fft_wopbs_cache
+    }
+}
+
+impl KeyCacheAccess for PackingKeySwitchTestParams<u32> {
+    type Keys = PackingKeySwitchKeys<u32>;
+
+    fn access(keycache: &KeyCache) -> &KeyCacheCoreImpl<Self, Self::Keys> {
+        &keycache.u32_pksk_cache
+    }
+}
+
+impl KeyCacheAccess for PackingKeySwitchTestParams<u64> {
+    type Keys = PackingKeySwitchKeys<u64>;
+
+    fn access(keycache: &KeyCache) -> &KeyCacheCoreImpl<Self, Self::Keys> {
+        &keycache.u64_pksk_cache
+    }
+}
+
+lazy_static! {
+    pub static ref KEY_CACHE: KeyCache = Default::default();
+}
+
+// This function populates keycache for core_crypto. This is very helpful when doing code coverage
+// since generating keys in that case is very time consuming.
+//#[cfg(not(feature = "__coverage"))]
```

Comment:
this comment needs to be removed, I think it is the thing slowing down coverage as we are re generating keys as we run tests.

Also I would gate it behind internal-keycache to avoid generating keys for "fun" when running tests without keycache